server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -0,0 +1,371 @@
+                indexMetadataFileName(indexMetadata)
+            );
+            final UploadedIndexMetadata uploadedIndexMetadata = new UploadedIndexMetadata(
+                indexMetadata.getIndex().getName(),
+                indexMetadata.getIndexUUID(),
+                indexMetadataKey
+            );
+            allUploadedIndexMetadata.add(uploadedIndexMetadata);
+        }
+        final ClusterMetadataMarker marker = uploadMarker(clusterState, allUploadedIndexMetadata, false);
+        final long durationMillis = relativeTimeMillisSupplier.getAsLong() - startTimeMillis;
+        if (durationMillis >= slowWriteLoggingThreshold.getMillis()) {
+            logger.warn(
+                "writing cluster state took [{}ms] which is above the warn threshold of [{}]; " + "wrote full state with [{}] indices",
+                durationMillis,
+                slowWriteLoggingThreshold,
+                allUploadedIndexMetadata.size()
+            );
+        } else {
+            // todo change to debug
+            logger.info(
+                "writing cluster state took [{}ms]; " + "wrote full state with [{}] indices",
+                durationMillis,
+                allUploadedIndexMetadata.size()
+            );
+        }
+        return marker;
+    }
+
+    /**
+     * This method uploads the diff between the previous cluster state and the current cluster state. The previous marker file is needed to create the new
+     * marker. The new marker file is created by using the unchanged metadata from the previous marker and the new metadata changes from the current cluster
+     * state.
+     *
+     * @return The uploaded ClusterMetadataMarker file
+     */
+    @Nullable
+    public ClusterMetadataMarker writeIncrementalMetadata(
+        ClusterState previousClusterState,
+        ClusterState clusterState,
+        ClusterMetadataMarker previousMarker
+    ) throws IOException {
+        final long startTimeMillis = relativeTimeMillisSupplier.getAsLong();
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert previousClusterState.metadata().coordinationMetadata().term() == clusterState.metadata().coordinationMetadata().term();
+        assert REMOTE_CLUSTER_STATE_ENABLED_SETTING.get(settings) == true : "Remote cluster state is not enabled";
+        final Map<String, Long> previousStateIndexMetadataVersionByName = new HashMap<>();
+        for (final IndexMetadata indexMetadata : previousClusterState.metadata().indices().values()) {
+            previousStateIndexMetadataVersionByName.put(indexMetadata.getIndex().getName(), indexMetadata.getVersion());
+        }
+
+        int numIndicesUpdated = 0;
+        int numIndicesUnchanged = 0;
+        final Map<String, ClusterMetadataMarker.UploadedIndexMetadata> allUploadedIndexMetadata = previousMarker.getIndices()
```

Comment:
can we refactor the logic to find the incremental diff to private method.

Reply:
This refactor will be done along with parallel upload PR: https://github.com/soosinha/OpenSearch/pull/6